File: solutions/clojure/11a.clj
(ns aoc2024.day11
  (:require [clojure.string :as str]))

;; ## Transition Function (to handle stone transitions)
(defn transition [stone]
  (cond
    (zero? stone) [1]  ;; If stone is 0, transition to [1]
    (even? (count (str stone)))  ;; If number of digits is even
    (let [s (str stone)
          len (count s)
          pivot (quot len 2)]  ;; Split the stone into two parts
      (map parse-long [(subs s 0 pivot) (subs s pivot len)])) 
    :else [(* 2024 stone)]))  ;; Else multiply by 2024

;; ## Part-1 Logic
(defn part-1 [s]
  (let [initial-stones (->> s
                            (mapcat #(re-seq #"\d+" %))  ;; Extract numbers from all lines
                            (map parse-long))  ;; Parse numbers into integers
        bs (iterate #(mapcat transition %) initial-stones)]  ;; Create an infinite sequence of transitions
    (-> bs (nth 25) count)))  ;; Count the number of stones after 25 iterations


;; ## Main function
(defn -main [& args]
  (let [filename (first args)  ;; Get the input file name
        lines (str/split-lines (slurp filename))]  ;; Read the input file
    (println (part-1 lines))))

File: solutions/python/11a.py
from collections import deque

NUM_BLINKS = 25

def count_digits(n):
    digits = 1
    while n >= 10:
        n //= 10
        digits += 1
    return digits

def update_arrangement(arrangement):
    new_arrangement = deque()
    while arrangement:
        stone = arrangement.popleft()
        if stone == 0:
            new_arrangement.append(1)
        else:
            digits = count_digits(stone)
            if digits % 2 == 0:
                half = digits // 2
                divisor = 10 ** half
                left = stone // divisor
                right = stone % divisor
                new_arrangement.append(left)
                new_arrangement.append(right)
            else:
                new_arrangement.append(stone * 2024)
    return new_arrangement

if __name__ == "__main__":
    with open("input.txt") as f:
        stones = deque(map(int, f.read().split()))

    for _ in range(NUM_BLINKS):
        stones = update_arrangement(stones)

    print(len(stones))
